src/query_plan/fetch_dependency_graph.rs

```diff
@@ -144,3 +146,162 @@ pub(crate) struct FetchDependencyGraphPath {
     response_path: Vec<FetchDataPathElement>,
 }
+
+#[derive(Debug, Default)]
+pub(crate) struct NodePath {
+    full_path: Vec<QueryPathElement>,
+    path_in_group: Vec<QueryPathElement>,
+    response_path: Vec<ResponsePathElement>,
+}
+
+#[derive(Debug)]
+pub(crate) struct DeferContext {
+    current_defer_ref: Option<NodeStr>,
+    path_to_defer_parent: Vec<QueryPathElement>,
+    active_defer_ref: Option<NodeStr>,
+    is_part_of_query: bool,
+}
+
+impl Default for DeferContext {
+    fn default() -> Self {
+        Self {
+            current_defer_ref: None,
+            path_to_defer_parent: Vec::new(),
+            active_defer_ref: None,
+            is_part_of_query: true,
+        }
+    }
+}
+
+impl FetchDependencyGraph {
+    pub(crate) fn new(
+        supergraph_schema: ValidFederationSchema,
+        federated_query_graph: Arc<QueryGraph>,
+        root_type_for_defer: Option<CompositeTypeDefinitionPosition>,
+        starting_id_generation: u64,
+    ) -> Self {
+        Self {
+            defer_tracking: DeferTracking::empty(&supergraph_schema, root_type_for_defer),
+            supergraph_schema,
+            federated_query_graph,
+            graph: Default::default(),
+            root_nodes_by_subgraph: Default::default(),
+            starting_id_generation,
+            fetch_id_generation: starting_id_generation,
+            is_reduced: false,
+            is_optimized: false,
+        }
+    }
+
+    /// Must be called every time the "shape" of the graph is modified
+    /// to know that the graph may not be minimal/optimized anymore.
+    fn on_modification(&mut self) {
+        self.is_reduced = false;
+        self.is_optimized = false;
+    }
+
+    pub(crate) fn get_or_create_root_node(
+        &mut self,
+        subgraph_name: &NodeStr,
+        subgraph_schema: &ValidFederationSchema,
+        root_kind: SchemaRootDefinitionKind,
+        parent_type: CompositeTypeDefinitionPosition,
+    ) -> NodeIndex {
+        self.root_nodes_by_subgraph
+            .get(subgraph_name)
+            .copied()
+            .unwrap_or_else(|| {
+                let node = self.new_node(
+                    subgraph_name.clone(),
+                    subgraph_schema.clone(),
+                    parent_type,
+                    /* has_inputs: */ false,
+                    root_kind,
+                );
+                self.root_nodes_by_subgraph
+                    .insert(subgraph_name.clone(), node);
+                node
+            })
+    }
+
+    pub(crate) fn new_node(
+        &mut self,
+        subgraph_name: NodeStr,
+        subgraph_schema: ValidFederationSchema,
+        parent_type: CompositeTypeDefinitionPosition,
+        has_inputs: bool,
+        root_kind: SchemaRootDefinitionKind,
+        // merge_at: Option<Vec<ResponsePathElement>>,
+        // defer_ref: Option<NodeStr>,
+    ) -> NodeIndex {
+        self.on_modification();
+        self.graph.add_node(Arc::new(FetchDependencyGraphNode {
```

Comment:
Since that’s exactly how the `subgraph_schema` argument was obtained when calling this, I’ve removed the argument and made this method call `schema_by_source` instead.

Reply:
Makes sense!